### PR TITLE
codeintel: Support multipart LSIF index uploads

### PR DESCRIPTION
**Pull request by efritz**
This implements [RFC 143: Multipart uploads](https://docs.google.com/document/d/1nN9mKnQ-5OSPnWg0mHAJ9T3p1jEdnDXgb3HgcnHNckQ).

Overview of changes:

- Bundle manager now has two new endpoints: one to upload a _partial_ index, and one to stitch together many partial indexes into a single compressed file.
- The api server now supports additional query parameters for its upload endpoint. These query parameters are supplied via src-cli (as of https://github.com/sourcegraph/src-cli/pull/197).

### CODE REVIEW

File: internal/codeintel/db/uploads.go [hunk 1]
```diff
@@ -147,22 +147,58 @@ func makeSearchCondition(term string) *sqlf.Query {
...
-			VALUES (%s, %s, %s, %s, %s, %s, %s)
+			INSERT INTO lsif_uploads (
+				commit,
+				root,
+				repository_id,
+				indexer,
+				state,
+				num_parts,
+				uploaded_parts,
+				tracing_context
```
Comment (gbrik) [inline internal/codeintel/db/uploads.go:167]: question:
what is the tracing_context field in the DB?

File: internal/codeintel/db/uploads_test.go [hunk 2]
```diff
@@ -190,14 +191,74 @@ func TestGetUploadsByRepo(t *testing.T) {
...
+		State:             "uploading",
+		FailureSummary:    nil,
+		FailureStacktrace: nil,
+		StartedAt:         nil,
+		FinishedAt:        nil,
+		TracingContext:    "{}",
+		RepositoryID:      50,
+		Indexer:           "lsif-go",
+		NumParts:          3,
+		UploadedParts:     nil,
```
Comment (gbrik) [inline internal/codeintel/db/uploads_test.go:229]: question:
don't we expect uploaded parts to be an empty array?

File: cmd/precise-code-intel-api-server/internal/server/enqueue.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+// requests for the same upload.
+func (s *Server) handleEnqueueMultipartSetup(r *http.Request, uploadArgs UploadArgs, numParts int) (interface{}, error) {
+	id, err := s.db.InsertUpload(r.Context(), &db.Upload{
+		Commit:        uploadArgs.Commit,
+		Root:          uploadArgs.Root,
+		RepositoryID:  uploadArgs.RepositoryID,
+		Indexer:       uploadArgs.Indexer,
+		State:         "uploading",
+		NumParts:      numParts,
+		UploadedParts: nil,
```
Comment (gbrik) [inline cmd/precise-code-intel-api-server/internal/server/enqueue.go:166]: suggestion:
can we set this to an empty array and enforce in the `InsertUpload` method that the field is non-nil?

File: cmd/precise-code-intel-api-server/internal/server/enqueue_test.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+			t.Fatalf("unexpected error reading payload: %s", err)
+		}
+
+		if diff := cmp.Diff(expectedContents, contents); diff != "" {
+			t.Errorf("unexpected file contents (-want +got):\n%s", diff)
+		}
+	}
+}
+
+func TestHandleEnqueueMultipartFinalize(t *testing.T) {
```
Comment (gbrik) [inline cmd/precise-code-intel-api-server/internal/server/enqueue_test.go:231]: request:
can we have a test for querying "done" when we haven't uploaded all the parts?

Comment (efritz) [reply cmd/precise-code-intel-api-server/internal/server/enqueue_test.go:231]: Yes! Please request other low-hanging tests I've missed when you see them.

File: internal/codeintel/db/uploads.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/db/uploads.go:167]: It's something we'll probably get rid of soon - when we were using typescript I serialized the opentracing context at the time of the upload creation to be pulled by the worker.

File: cmd/precise-code-intel-bundle-manager/internal/server/stitch.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/inconshreveable/log15"
+	"github.com/sourcegraph/sourcegraph/cmd/precise-code-intel-bundle-manager/internal/paths"
+)
+
+// stitchMultipart combines multiple uploads into a single compressed file. Each part on disk
+// with the name `{bundleId}.{partIndex}.lsif.gz` will be concatenated into a single bundle
+// file with the name `bundleId.lsif.gz`. The content of each part is decompressed and written
+// to a new compressed file sequentially. On success, the part files are removed.
```
[the remark this answers is not in the document]
Comment (efritz) [reply cmd/precise-code-intel-bundle-manager/internal/server/stitch.go:16]: If we'd like to move to being able to process things as they are being uploaded I think it's easier to do it the way that's currently implemented, no? I may be wrong in that assumption though.

File: cmd/precise-code-intel-api-server/internal/server/enqueue.go (hunk 3 above)
Comment (efritz) [reply cmd/precise-code-intel-api-server/internal/server/enqueue.go:198]: Updated (the issue was in the scan procedure).

File: internal/codeintel/db/uploads_test.go (hunk 2 above)
Comment (efritz) [reply internal/codeintel/db/uploads_test.go]: Updated (the issue was in the scan procedure).

File: cmd/precise-code-intel-api-server/internal/server/enqueue_test.go (hunk 4 above)
Comment (efritz) [reply cmd/precise-code-intel-api-server/internal/server/enqueue_test.go:298]: Added one.

File: cmd/precise-code-intel-bundle-manager/internal/server/stitch.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gbrik) [reply cmd/precise-code-intel-bundle-manager/internal/server/stitch.go:16]: ooh i didn't think about that, good point

**Pull request merged**